### PR TITLE
Add a spyglass lensserver and a trival stepgraph lens

### PR DESCRIPTION
**Pull request by alvaroaleman**
This PR is mostly intended to provide the boilerplate to build a
lensserver and add the first parts of a simple step graph lens.

### CODE REVIEW

File: cmd/lensserver/README.md [hunk 1]
```diff
@@ -0,0 +1,9 @@
+# Lensserver
+
+This binary provides additional lenses for Prows spyglass log viewer.
```
Comment (petr-muller) [inline cmd/lensserver/README.md:3]: nit: "lens" is already Prow jargon, I think a link to https://github.com/kubernetes/test-infra/blob/master/prow/spyglass/write-a-lens.md and/or https://github.com/kubernetes/test-infra/blob/master/prow/spyglass/architecture.md#spyglass-lenses would be useful here.

File: cmd/lensserver/main.go [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+	logrus.SetFormatter(&logrus.JSONFormatter{})
+
+	configAgent := &config.Agent{}
+	if err := configAgent.Start(o.configPath, o.jobConfigPath); err != nil {
+		logrus.WithError(err).Fatal("Error starting config agent.")
+	}
+
+	logLevel, err := logrus.ParseLevel(configAgent.Config().LogLevel)
+	if err != nil {
+		logrus.Fatalf("Could not parse loglevel %q: %v", configAgent.Config().LogLevel, err)
```
Comment (petr-muller) [inline cmd/lensserver/main.go:58]: nit: in other `Fatal`s you're using `WithError`, why not here?

File: cmd/lensserver/main.go [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+		Config: common.LensOpt{
+			LensName:  lens.Config().Name,
+			LensTitle: lens.Config().Title,
+		},
+		Lens: lens,
+	}}
+
+	lensServer, err := common.NewLensServer(spyglassLocalLensListenerAddr, ja, spyglass.NewStorageArtifactFetcher(opener, configAgent.Config, false), spyglass.NewPodLogArtifactFetcher(ja), configAgent.Config, localLenses)
+	if err != nil {
+		logrus.Fatalf("Failed to start lens server: %v", err)
```
Comment (petr-muller) [inline cmd/lensserver/main.go:81]: `WithError`

File: hack/run-lens.sh [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+set -euo pipefail
+
+which deck &>/dev/null || k8s.io/test-infra/prow/cmd/deck
+
+export KUBECONFIG="$(mktemp)"
+
+cleanup(){
+  rm -f $KUBECONFIG
+  rm -f $deck_config
```
Comment (petr-muller) [inline hack/run-lens.sh:11]: Plz run this through shellcheck and quote things :)

**Review by petr-muller — COMMENTED** [review]
few code nits, but I'd like the links in the readme

File: pkg/lenses/stepgraph/stepgraph.go [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+// Header renders the content of <head> from template.html.
+func (lens Lens) Header(artifacts []api.Artifact, _ string, config json.RawMessage) string {
+	var buf bytes.Buffer
+	if err := tmpl.ExecuteTemplate(&buf, "header", nil); err != nil {
+		return fmt.Sprintf("<!-- FAILED EXECUTING HEADER TEMPLATE: %v -->", err)
+	}
+	return buf.String()
+}
+
+// Callback does nothing.
```
Comment (stevekuznetsov) [inline pkg/lenses/stepgraph/stepgraph.go:56]: Explain why? This is to handle click-throughs inside of lenses, right?

File: pkg/lenses/stepgraph/stepgraph.go [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+func (lens Lens) Callback(artifacts []api.Artifact, resourceDir string, data string, config json.RawMessage) string {
+	return ""
+}
+
+// Body renders the <body>
+func (lens Lens) Body(artifacts []api.Artifact, resourceDir string, data string, config json.RawMessage) string {
+	if len(artifacts) != 1 {
+		logrus.WithField("artifacts_count", fmt.Sprintf("%d", len(artifacts))).Error("Expected exactly one artifact")
```
Comment (stevekuznetsov) [inline pkg/lenses/stepgraph/stepgraph.go:64]: Why the `fmt.Sprintf` here?

**Review by stevekuznetsov — COMMENTED** [review]
Any screenshots?

File: hack/run-lens.sh (hunk 4 above)
Comment (alvaroaleman) [reply hack/run-lens.sh:11]: If anything of these ever contains something that would need quoting the whole thing will break anyways

File: cmd/lensserver/main.go (hunk 3 above)
Comment (alvaroaleman) [reply cmd/lensserver/main.go:81]: I blame copy pasta

File: pkg/lenses/stepgraph/stepgraph.go (hunk 5 above)
Comment (alvaroaleman) [reply pkg/lenses/stepgraph/stepgraph.go:56]: This is to handle fetching additional data after the page was already loaded to not require all data during initial load. The only upstream lens where this does something is the buildlog lens, but it's part of the interface so it needs to exist.

File: hack/run-lens.sh (hunk 4 above)
Comment (petr-muller) [reply hack/run-lens.sh]: I will not block the PR on this, but in general quoting is bash good form, and it allows using tools like shellcheck to spot more important errors without drowning in warnings.

**Comment by alvaroaleman** [comment]
> Any screenshots?

It's the same as what I showed in the call the other day:

collapsed
less_collapsed
expanded

**Comment by alvaroaleman** [comment]
Part of https://issues.redhat.com/browse/DPTP-1507

**Comment by stevekuznetsov** [comment]
/lgtm
/hold

@petr-muller yours to unhold

**Pull request merged**